apps/reaper/lib/reaper/xml/xml_stream/sax_handler.ex

```diff
@@ -0,0 +1,133 @@
+defmodule XMLStream.SaxHandler do
+  @moduledoc false
+
+  defmodule State do
+    @moduledoc false
+    @enforce_keys [:tag_path, :emitter]
+    defstruct stack: [], accumulate: false, tag_stack: [], tag_path: [], emitter: nil
+
+    def new(fields \\ %{}) do
+      struct!(__MODULE__, fields)
+    end
+
+    def update(%__MODULE__{}=state, fields) do
+      struct!(state, fields)
+    end
+  end
+
+  @behaviour Saxy.Handler
+  @file_chunk_size 40_000
+
+  ######################
+  ## Client Functions ##
+  ######################
+  def start_stream(filepath, tag_path_string, callback) when is_function(callback, 1) do
+    tag_path = parse_selector(tag_path_string)
+
+    filepath
+    |> File.stream!([], @file_chunk_size)
+    |> handle_ufeff()
+    |> Saxy.parse_stream(
+      __MODULE__,
+      __MODULE__.State.new(tag_path: tag_path, emitter: callback)
+    )
+  end
+
+  ###############
+  ## Callbacks ##
+  ###############
+  def handle_event(:start_document, _, %State{}=state) do
+    ok(state)
+  end
+
+  def handle_event(:end_document, _, %State{}=state) do
+    ok(state)
+  end
+
+  def handle_event(:start_element, {tag_name, attributes}, %State{}=state) do
+    state = State.update(state, tag_stack: [tag_name | state.tag_stack])
+
+    if state.tag_stack == state.tag_path or state.accumulate do
+      tag = {tag_name, attributes, []}
+
+      state
+      |> State.update(stack: [tag | state.stack], accumulate: true)
+      |> ok()
+    else
+      ok(state)
+    end
+  end
+
+  def handle_event(:characters, chars, %State{stack: stack, accumulate: accumulate} = state) do
+    if accumulate do
+      [{tag_name, attributes, content} | stack] = stack
+
+      current = {tag_name, attributes, [chars | content]}
+
+      state
+      |> State.update(stack: [current | stack])
+      |> ok()
+    else
+      ok(state)
+    end
+  end
+
+  def handle_event(
+        :end_element,
+        tag_name,
+        %State{stack: stack, accumulate: accumulate, tag_stack: [_stack_tag_name | tag_stack]} = state
+      ) do
+    state = %{state | tag_stack: tag_stack}
+
+    if accumulate do
```

Comment:
Again, I think you can pattern out this `if/else`.

Reply:
This one I'm not sure about. In this one the function head is popping the `_stack_tag_name` off the list and then updating the state before the `if` block.

We could pattern match the accumulate but then we'd have to repeat that logic in two functions/heads, or create a function to share that functionality. Both options would work but they seem more indirect/complicated than just using an `if` in this case.

What are your thoughts?

Reply:
I see what you're saying. This function is really big though, so it should probably be broken down anyway. How about something like:

```elixir
def handle_event(:end_element, tag_name, %State{tag_stack: [_ | tag_stack]} = state), do: private_function(tag_name, %{state | tag_stack: tag_stack})

defp private_function(tag, %State{accumulate: true} = state) do
  # if accumulate true logic
end

defp private_function(_tag, state), do: ok(state)
```

Obviously come up with a better private function name 😛

Reply:
I did a big refactor of the handler that should cover this